### ISSUE TITLE
Would it be possible to make Coord immutable, to be able to cache instances ?

### ISSUE THREAD

**Issue by smelc** [SquidPony/SquidLib#73]
Because I work in a safety-focused company (where we prove that software is defect-free), I like [immutable](https://en.wikipedia.org/wiki/Immutable_object) objects as they allow safe sharing. That is why I would have written `Coord` in an immutable way (i.e. with final fields, and with mutating methods that return another object). Another benefit is that it allows to use a cache such as this one (which I use in my game):

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.java}
package hgames.shared

import squidpony.squidmath.Coord;

/**
 * Pre-allocated instances of {@link Coord}. You should not mute them, as they
 * are shared.
 * 
 * @author smelc
 */
public class Coords {

	/** The pool whose content is lazily allocated. */
	private static final Coord[][] POOL;

	private static final int POOL_BOUND = 256;

	static {
		POOL = new Coord[POOL_BOUND][POOL_BOUND];
	}

	/**
	 * @param x
	 * @param y
	 * @return A {@link Coord} at coordinates {@code x} and {@code y}
	 */
	/*
	 * If this method is used in a multithreaded environment, there are races on
	 * the content of {@link #POOL}. But they are benign, because all threads
	 * put equivalents value: we don't care which tread's value makes it to
	 * #POOL. The only harm is that we may allocate more than once a Coord, but
	 * I prefer that over synchronizing this method.
	 */
	public static Coord get(int x, int y) {
		if (0 <= x && x < POOL_BOUND && 0 <= y && y < POOL_BOUND) {
			Coord c = POOL[x][y];
			if (c == null) {
				c = new Coord(x, y);
				POOL[x][y] = c;
			}
			assert c.x == x && c.y == y : "The cache of coordinates got corrupted";
			return c;
		} else
			return new Coord(x, y);
	}
}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This cache would save a lot of allocations in many parts of SquidLib (for example in path finding algorithms), but it would force containers of Coords (for example monsters or players) to have a mutable Coord field. I think this should not be an issue as SquidLib has a pretty mutable style for now. I've checked callers of mutating methods of Coord (in squidlib and squidlib-util, but not in demos), and it seems to me the change is doable everywhere. In particular, squidlib never mutates a Coord given by a client (which wouldn't be very nice to him !), all Coords that are mutated are allocated within squidlib itself. Of course, I can take care of doing the change.

**Comment by davidbecker** [comment]
I like the idea of immutable objects very much my self, but I think the fixed size pool for all coordinates is an arbitrary limit that limits the usefulness in a general library.
But this is only a minor implementation detail - I like the general idea though :)

**Comment by tommyettinger** [comment]
I actually made this change once before, when I ported SquidLib 1.95's logic code out into its own library, Cuttlebone. Cuttlebone didn't have a fixed-size pool and I think it isn't a great idea for such tiny objects. I definitely am in favor of immutable data where possible; I use Clojure a lot but it suffers from some performance trouble that is only partly a result of immutability by default (it has bigger problems from boxed math). I believe all Coords that are used by or can be produced by SquidLib are rather closely tied to map size (an exception is Coord3D, which SColor uses as 0-255 RGB in place of XYZ for a line that is used as a gradient generator). It can be safely assumed, correct me if I'm wrong, that maps will be no more than 1000x1000, since DijkstraMap and dungeon generation in general slow to a crawl with too many cells. If we do have a hard limit then 2^15, with over 32,000 values, is clearly safe -- and it means the x or y coordinate fits in a `short`, which means both can fit in an `int` or a Coord3D in a `long` using a simple `<<` and `|`. What are your thoughts on a (possibly alternative) Coord API that doesn't have objects, just static methods on Coord that return primitive numbers?
```java
int pt = Coord.coord(3, 2); // pt is now the int 0x30002
Coord.x(pt); // returns 3
Coord.y(pt); // returns 2
pt += Coord.X1; // X1 is a static final int corresponding to 0x10000;
pt -= Coord.Y1; // Y1 is similar, but is just 1, so now...
Coord.x(pt); // returns 4
Coord.y(pt); // returns 1
System.out.println(Integer.toHexString(pt)); // prints "40001", toHexString omits the "0x"
System.out.println(Coord.toString(pt)); // prints "x: 4, y: 1"
```

This has a few key advantages, namely it wouldn't use a pool, and numbers are about as immutable as you can get, not being reference types at all! We also get to use operators directly on our "Coord objects, not really" and they behave like you'd expect, except in the case of negative x or y. Saving on allocations probably helps with GC performance as well, particularly on Android. Downsides are that it could be hard to tell which ints are supposed to represent Coords and which aren't, including at an API level. There are a lot of functions that already have a version for Coords and a version for separate x, y parameters, but there are still more that have only one or the other. This is just an idea at this point and I don't know the precise performance ramifications.

**Comment by smelc** [comment]
@tommyettinger> Your proposal is smart, and I bet it'd produce the fastest code. But, as you say, it would make the code harder to understand, pretty much everywhere :-/, so I'd avoid such a change as long as it isn't _required_ for performances.

Maybe my pool is oversized, indeed my maps are at most of size 100x50 :smile: and the pool is only partly shallow: all arrays are initialized at creation. So would the following changes make it more appealing to you:

* Reduce the pool size, so that it'll almost always make sense (32 or 64 ?), no matter the application. After all, if the pool is not big enough, that's no big deal, a new Coord is allocated every time, which pretty much defaults to the existing behavior.
* Make the pool really shallow, by only allocating one array initially and initializing the columns on demand. This would make benign races (see my comment in the pool's code) happen more often, but I'm confident it'd be harmless (we have some code with such benign races at work, and it did not cause problems).

?

I'd go for the first option, as it is the safest one, with a pool of size 128. Regarding memory consumption, such a pool is only 64K on a 32-bits machine, so that the size I'd choose (but I have no knowledge of Android's constraints).

**Comment by tommyettinger** [comment]
I've thought about this pretty much all day today. I looked into annotation processors as a way to effectively provide a type alias for Coord to make it act like an int, but no matter how well or poorly that would work, it would still need to be boxed to fit into practically any collection, and that takes away a lot of the benefit. The original idea is still the best option, taking into account @davidbecker 's configurable-size-pools recommendation.
So, I've changed my mind on the pool idea, as long as the library user has a way to configure its size -- being able to consistently refer to the same immutable Coords throughout the entirety of SquidLib should be a significant performance aid by avoiding GC and re-allocation for numerous small objects. I realize now just how many Coords I create in the process of AOE calculations, and even though performance is fine on a high-end recent laptop, the same might not be true for a low-end out-of-date Android tablet.
Because Coords are (going to be) immutable, as long as they're always initialized when the program starts, they can safely be be made static, even on Android, since lingering immutable values just mean less work to initialize and result in the same values. This means there can be a single static pool for all Coords within a certain set of bounds.
So here's what I propose: I'll make `Coord` immutable and write the pooling inside `Coord`, using @smelc 's code as the core. `Coord`'s pool will have a static 2D array of immutable `Coord`s (probably 256x256 because that's more than enough for most games), and a pair of methods to expand the pool and fill it with the appropriate positive values. Expanding should be OK, but contracting could cause problems if the pool lingers on Android, and wouldn't help a whole lot with memory consumption since as was pointed out, the whole array is far less than one megabyte anyway. Negative values for x and y are used occasionally, but I believe they never go below -1 in the library as it stands, and will probably go no lower than -2 in a future version when I add thin wall support. At least a small amount of the 2D array will be offset into negative x and y, probably with `POOL[0][0] = new Coord(-3, -3);` and `POOL[255][255] = new Coord(252, 252)` to ensure a 250 height, 250 width map could be represented with room to spare. The core 256x256 and directions will be initialized in a static{} block because it sacrifices a tiny bit of startup time and a fraction of RAM to avoid additional checks at runtime every time the array is accessed. I'm thinking the constructor for `Coord` will be protected, and access to `Coord`s will be done through a static method that prefers looking up in the pool and only constructs a new immutable Coord if its x or y are out of bounds.
Whew! I don't know how much of that anyone felt like reading, but... I like the talent SquidLib is attracting :+1: Thanks for the firm push and solid code, @smelc , and thanks for the feedback, @davidbecker !

**Comment by tommyettinger** [comment]
This is done and it seems to have no behavior differences or noticeable slowdown in EverythingDemo. I'll run some more tests, but I had to uh... change a lot of APIs. Most importantly, you probably need to do a find/replace in any code using SquidLib master, on `new Coord(` replacing with `Coord.get(` , which should make any instantiations of Coord avoid using the now-non-visible constructor directly. Anything that sets a Coord's x or y no longer mutates it in place, and now returns a reusable Coord from the pool with the desired x or y change and the other one kept as-is. You can set x and y simultaneously with the static Coord.get or "modify" their current values on an object with the method on Coord, `translate(int xMod, int yMod)`, which returns a different Coord from the pool as well. Expect a lot of assignment of one immutable Coord to a variable that was holding another Coord, which thankfully involves no copying!
Leaving this open until someone else can verify that this works.

**Comment by smelc** [comment]
I felt like reading to the end :smile: Thank you for doing it, in my experience, immutability protects against a lot of mistakes! And allocation was always the bottleneck (I mean: once the right algorithm has been chosen) in computing-intensive Java-stuff I do at work, so even if I won't neither be able to notice the difference on my development machine; I'm very confident it'll help on old systems. I'll confirm it's working when I've updated to the tip and tested on my game.

**Comment by davidbecker** [comment]
@tommyettinger I also felt like reading to the end :wink:
I have an old (2007) lapto that I use as a performance baseline, I will test your changes there when I'm back from work and will report back how your changes affected the performance.

**Comment by davidbecker** [comment]
TL:DR the new Coord implementation is faster

now the long version :smiley:
I wrote a test that did some FOV & LOS computation over a small map (30*30).
This test was run 100 time on 4 threads on my old laptop (core2duo mashine) and the amount that each thread took was noted.
![coord](https://cloud.githubusercontent.com/assets/1927731/9980377/9e046d30-5f97-11e5-9dd9-f435746040bb.png)
The position of each dot is the time that it took for this particular thread to finish. The size of the circle is the absolute differenze between the time of the single test and the arithmatic mean of the whole series. The last two tests in each series shouldn't count, since the mashine isn't at its limit anymore...

Anyway:
The arithmatic mean for the old implementation was 91625,81 ms (including the last two tests), and the mean for the new implementation was 90840,53 ms - so the new implementation was overall faster, even the last rounds couldn't change it.

I haven't checked the amount of gc events that are generated by each implementation, this would be very interesting as well

**Comment by smelc** [comment]
@davidbecker > That's interesting data. Clearly, allocation of `Coord` isn't the bottleneck in this test, otherwise the difference would be larger.

**Comment by davidbecker** [comment]
@smelc you are probably right about that about that. Maybe it is the memory allocation? I haven't had time to look at the new implementation of Coord myself, but maybe there is room for improvement. I'm sure there is room for performance improvements in the implementation of the algorithms as well.
I stumbled upon an interesting youtube series from the google android developers a couple of month ago that is focused on improving performance on android applications. Some parts are android specific (the gc  in particular), but most of the stuff is true for general java as well. It is for the most part obvious stuff, but maybe you can learn a thing or two...
Have a look at the [playlist](https://www.youtube.com/playlist?list=PLOU2XLYxmsIKEOXh5TwZEv89aofHzNCiu) when you have some time :smiley:

**Comment by tommyettinger** [comment]
LOS using Elias and I suspect some FOV for large radii are slow for other reasons. Elias is likely the reason for the biggest speedup because it probably takes the most time, and because it creates a lot of Coords, one for every cell in the line and then some (cells immediately adjacent to the Elias line also may be part of the line it returns). I would be curious to see this test without Elias (to see if the most-Coord-heavy test is responsible for the most speedup) or with some DijkstraMap usage (to test another common source of slowdown when many monsters need pathfinding at once). DijkstraMap definitely will be re-using at least some Coords with the new version, though it primarily depends on 2D double arrays and manipulates the dungeon level's data all at once. I can write the test, but not run it under the same circumstances.

**Comment by davidbecker** [comment]
@tommyettinger are you sure the pool works as intended? To be honest I can't figure out some of the implementation details; the "-3" in the get() method, is it to have some buffer on the edges? I think the problem is in the get() method itself. If the parameters aren't describing a Coord in the current pool a new Coord is returned on line 26, but as far as I can see it isn't added to the pool - when get() is called with the same parameters a new object should be created each time. It appears to me that the pool should be increased instead and a Coord returned from the increased pool.

But I think that at the and of the day we need to rewrite some labor intensive algorithms to work with your excelent proposal to realy gain a performance boost. We could even use the 16 bit of a short to represent a full coordinate on a small 2D map (2^8=256 possible positions for x and y each).
But for this to be manageable we would need automated test coverage for all the algorithms that should be rewritten in a framework like JUnit to make sure that it won't blow up in our face :wink: The current System.out stile test are a little too fragile for me to be onest

**Comment by tommyettinger** [comment]
Yeah, the pool currently doesn't cache odd-man-out Coords that are beyond the reach of the 2D array, but it does provide a way to expand (not ever contract) the 2D array to beyond the current size, initializing it with Coords so any access into the array will access a cached object. The -3 is so the 256-wide, 256-high array contains points from (x:-3,y:-3) to (x:252,y:252), ensuring 0-249 are valid for maps up to 250 length in either dimension and ensuring the few usages of small negative coords get cached as well.  Coord.get(0,0) returns POOL[3][3], which contains the cached Coord (x:0,y:0).  Coord.get(-3,-3) is also valid, and returns POOL[0][0], which contains the cached Coord (x:-3,y:-3).  Coord.get(-999,-3000) is outside the bounds of the 2D array POOL, so it returns a single uncached Coord (x:-999,y:-3000). We could make an attempt to cache the odd outliers but they should be incredibly rare within SquidLib, if they appear at all, and the expanding mechanism is there for games that need larger maps and can tolerate a larger pool.

I think what we need to do before we do any more large rewrites is to *profile as many of the performance tests* as we can, and probably to write more performance tests. JUnit should be handy here; even though I've never used it, I am aware that it is the standard. From my anecdotal experience on trying to optimize the technique processing, the slowest algorithms are LOS-related, particularly whenever Elias is used. I tried removing LOS checks which determined if a technique checker should process a cell, and simply _processing every cell_ turned out faster. (I think I added a non-LOS-related cutoff to skip over unlikely targets anyway.) Elias is more complex than it seems, it actually draws one Bresenham line and then draws dozens of additional Bresenham lines to check how anti-aliased the line should be on the fringes. The LOS class also uses multi-threaded code for Elias only to attempt a speedup, and it still doesn't help much vs. other LOS strategies.

Possibilities for speedup other than Coord caching:
  * A new (better) LOS method, probably tweaking one of the methods presented on [Roguebasin](http://www.roguebasin.com/).
  * Optional precomputation of difficult, reproducible results for things like LOS and possibly FOV.
    * This could be run on a second thread while the player plays the first level, and only await a result (causing a loading screen) if the precomputation hasn't finished by the time they get to the next level.
  * Dungeon generation with DungeonGenerator or (DungeonBoneGen, which the former uses) draws from one 20 sets of predefined chunks of map. These chunks are either 20x10 or 10x20 cells in size, and could have some information precomputed about what blocks LOS and FOV within the chunk. Precomputation won't work for some kinds of map currently; DungeonGenerator can modify the map when it adds doors (and walls to attach the doors to), and rather than checking every gap on the map for door-placement-viability, it might be suitable to push door-can-go-here into the chunks as well.
  * Better data structures would be handy.
    * [R-trees](https://en.wikipedia.org/wiki/R-tree) could be used for spatial lookups on sparse-Coord-key collections (possibly) more efficiently than HashMaps with Coord keys.
    * Shadowcasting FOV and other methods that produce or process a 2D 1-or-0 array could potentially be massively compressed by using a [Hilbert Curve](https://en.wikipedia.org/wiki/Hilbert_curve) or [Moore Curve](https://en.wikipedia.org/wiki/Moore_curve) or other way to draw a path across every cell of the dungeon. To generate a compressed FOV field as a list of integers: Starting at the minimum distance from the start of the Curve, mark the number of contiguous cells along the Curve that cannot be seen as a negative number, then the number of contiguous cells that can be seen as a positive number, then if more cannot be seen at a later point on the curve, another negative number for those, another number for cells after that that can be seen, etc. The minimum size for this encoding is 3 `int`s and whatever the memory cost of a 1d array is, the first `int` being the starting corner or out-of-bounds Coord, the second being the player seeing his own square and the third being all other squares being unseen. Plus, assuming a way to convert from a 2D Coord to position on a Hilbert Curve efficiently, you can tell whether a range of cells are in FOV relatively efficiently, and convert back to a 2D array (less efficient) or iterate down the Hilbert Curve in large hops (very efficient) to render the FOV. If we limit the absolute maximum size of a map to 250 height and 250 width (which seems reasonable for most games and the algorithms/pooling we have), then the Curve can be precomputed as a constant and we can always use the 256x256 version of the Curve. That also means a position on the Curve (and thus any Coord with non-negative x and y) can be stored as a 16-bit short, a straight line can be stored with 32 bits as an int, and all of these numbers can be used as indices into some 1D array that shares the Coord pool.
    * A method of having straight-line connectivity between all cells already done would be handy, but requires a relatively large amount of space; for a 60x60 map, the naive connect-all-to-all approach requires a 60x60x60x60 boolean 4D array, which is roughly 1.5MB for the booleans but at least that much again in just array references and length variables. For a 250x250 map, it's roughly half a gigabyte in just connection data, plus it would take forever to compute. It's also largely the same data; most cells can't see most other cells in the average dungeon. The previous compression method could help a lot here.

If someone is willing to step up and write a whole bunch of performance-related JUnit tests, I would strongly encourage @SquidPony to make that person a contributor (maybe I can, not sure), since it's unlikely to cause overlap with the last few parts of the code I'm working on before 3.0.0, and since it would mean less PR approval/merging getting in the way of test coverage (if we're talking about just writing tests, I can't see any reason to _not_ approve a pull request that adds a test).

**Comment by davidbecker** [comment]
We use JUnit at my workplace for test coverage and I would be willing to contribute testcases.
On the other hand I'm not really familiar with the library, since I only have used the dungeon generation and LOS in my own prototype. But this is a good thing for writing test cases :smiley:

**Comment by tommyettinger** [comment]
I have no idea where to post this, but I implemented Hilbert curves with run-length encoding for compressing FOV maps, and the results are... very nice.
```
Compression (Appropriate): 0.09112210361884926%
Compression (Approaching Worst-Case): 2.710843373493976%
```
Yes, that's smaller than 1/10 of one percent the size as a boolean[][] in optimal cases. And FOV currently produces double[][] FOV maps, so consider that too, haha.

(Approaching Worst-Case) is not actually the worst-case scenario; that would be a checkerboard of alternating true/false cells (and that better not be produced by a working FOV algorithm). This scenario does hit one of the weaknesses of the Hilbert curve; it uses a map that is drastically taller than it is wide, so visible cells could be in a region in the Hilbert curve where to get to an adjacent cell you need to walk through many non-visible cells, walk into maybe 1 or 2 other visible cells, then back into non-visible cells, repeatedly. (Appropriate) is the ideal case for the compression's memory usage; a large map with a relatively small percentage visible. Smaller maps are often significantly faster to process; a 120x120 map takes 1/4 the number of loops through the Hilbert indices, and a 60x60 map takes 1/16.

This should make caching FOV results much more palatable.

**Comment by tommyettinger** [comment]
@davidbecker I just re-read this issue thread and realized I should have rewritten a version of the performance test case to test Dijkstra Maps and another version to test AOEs with the new Coord setup. I'll get on it now.

Parallelization should offer impressive gains (for whatever we figure out is the biggest bottleneck) on desktop/laptop machines, but I have doubts about how effective it could be on Android devices (mostly older ones; I think the newest Samsung Galaxy S and Note phones and phablets have octacore processors, and those are fairly common, plus there's a good amount of surprisingly-low-price octacore tablets that can be bought from all sorts of companies. I don't know how efficient Android's Dalvik or ART VMs are at handling 8 or more threads, though...). One exception to these doubts is in caching FOV maps, which on the surface appears to be embarrassingly parallel and should benefit from even an old dual-core phone processor.

Also, a quick change that I should stick in a wiki: the cached Coord pool currently caches x and y between (inclusive) -3 and 255. This ensures the Hilbert curve can use Coords for the full size of 0-255 and also store Coords for the few negative x/y positions in use, though it rarely needs to use Coords for the Hilbert curve since it usually uses one lookup table for x at a given Hilbert distance and a separate one for y. The lookup tables are not completely insignificant in size; each is 128KB before JAR or APK compression (admittedly tiny compared to even one sound effect), and I'm not sure whether they should be calculated at runtime on-demand instead of being read from a resource stream (in fact, if they need to be unzipped from a JAR, it may be faster to calculate them than to read from a slow disk...).

**Comment by tommyettinger** [comment]
[Roadmap/idea page on the wiki](https://github.com/SquidPony/SquidLib/wiki/SquidLib-3.0-Changes). I'm still working on it.

**Comment by davidbecker** [comment]
@tommyettinger This sounds really promising - I experimented with caching computed FOV maps myself (nothing fancy, just a basic HashMap<Coord, double[][]> for each position that I computed the FOV before), which is a huge waste of memory in favor of computing time, but using Hilbert curves seems to make much more sense (I'm not pretending to realy know what these are, I just skimmed the wikipedia page 5 minutes ago..). If I understand it correctly you basically sort the FOV map after the lumination level and throw away the not lit part afterwards?

I will have a look at your test in 1032150e0a14fd07db4d9210fe09f322a32dc4d4 and will report back

**Comment by davidbecker** [comment]
@tommyettinger I ran the DijkstraPerformanceTest, here are the results for future references:
```
############################################################
#........#.........##........####....###...................#
#........#...............#...#####..##.....................#
#........................#...#####..##....###############..#
#...................#....#########..##....##...............#
#................####....##...............##...............#
#..................##.....................##..###########..#
#.....#....#.......##........####.........##..#............#
#.....#....#.......##........###############..#............#
###...######.....#########..################..######..######
##.....#####......########..##############.....#####.....###
#........##.......#####......#############.......##........#
#.........#....#...###...........#..########.....##........#
#...##.............##............#..#......##....##........#
#..####............##...##.......#..#.......##...#.........#
#..####......#.....##...#........#...........##..#.........#
#...##.........#...##.........................#............#
#.................###....#...##...............#.........#..#
##......###......#####.......##.......####....#..##.....#..#
###....#######..######......########..#####......####...####
############.....#####......#####.....#####......#####..####
###########........###..........#.....####....#..#####..####
#.#########........#######......##............#..#####..#..#
#..#####.....##....#............##...........##..#...#..#..#
#..####.......#....#........#########........##..#...#..#..#
#...##........#..###...######...............##.......#..#..#
#.............#.............#..............###.............#
#.............#.............#.....###########....#.........#
#........######..#####......#.....###########....###########
###.....#######....###.....#####..############..####.......#
###.....###...#....###.....#####......########..####.......#
#........##...#....#####..######.............#..####.......#
#.............#....####....#####.............#.....#.......#
#..##..#......#########....#####.........#...#.....#.......#
#.##...#..#...##########..######....######...#..##.#..######
#..##..#..#...............######....#....#####..##.........#
#.##......#.........................#....#......##.........#
#........##.......#####.............#...........####.......#
#........##.......########..######..#...........####.......#
##......########..#####......#####..######......#####.....##
##..############..#####......###.....#####......########..##
#....########......####......##........###..#####......#..##
#.....######..........#......##........###..#..........#...#
#......####...........########...##....###.........#...#...#
#.......##....###...........##....#....###......#..#.......#
#.##....##....####................#..#####..#####..#.......#
#.##....##....#####...#...........#...........#....#...#...#
####...####...#########....###....#...........#....#####..##
####...####...#########....########..#######..#....#......##
####..######..#########....########....#####..######......##
##.....#####..############..###...#....#####..######..######
#........###................###...#....#####..######..######
#........#.##.....................#....#####..#...##..######
#..##....#..######..######........######...#..#.......######
#...#....#...#####..######....#...######...#..#.......######
#...#..###.............####...#............#..#.......######
#...#....................###..#.........................####
#...#...........#####.....#####.......##................####
#####..#######..######.....####.......################..####
############################################################

invoking 100 tasks on 8 threads
30070
30419
30731
31403
29726
30687
29407
30279
23408
22335
22331
23398
23665
21828
23940
24050
24074
23891
24245
22736
22934
24497
22690
23159
24748
24270
23385
25586
25362
24450
25340
24030
24084
23839
24297
24237
24231
24212
23329
23529
22525
23930
23812
21909
21969
23464
22042
23239
22362
22184
22577
22828
22566
22233
21998
22404
20763
23402
23273
23850
23206
23325
23342
22981
24399
25044
24674
25361
25394
25067
24765
24637
21955
21424
22752
22087
22630
22699
22590
22464
22115
21800
22457
23004
22110
22174
22063
21769
21991
22144
23372
22752
22417
21720
22918
22136
14580
12056
10095
9824
```
I think that we could get the most performance gains on the lower end of the android spectrum (and the lower end of the desktop market as well) by focusing on making the computing intensive classes thread-safe by removing the need to store state in the class itself and refactor the calculation methods to work as static utility methods instead. This way we could compute different tasks in parallel without having the overhead of assigning new objects for each thread.
But this is just my gut feeling and I don't have any data to back it up, so take it for what it's worth

**Comment by tommyettinger** [comment]
Hoo boy, that is too many milliseconds for heavy DijkstraMap usage. On your machine, it looks to be about 4-5ms per Dijkstra scan (assuming about 5000-6000 calls are made to `scan()`, because it only calls it for walkable cells in the 60x60 dungeon, but calls it twice per test iteration). On my machine, with a quad-core 4th-gen i7 mobile processor, it was less than 1ms and closer to 1/2 or 2/3 of a ms. We can safely assume Android devices are often significantly worse than your laptop. I looked into the concurrent version of Dijkstra's algorithm, and it is only effective with a decent GPU. Using the GPU for computation on Android, possibly iOS via RoboVM, and desktop seems borderline impossible with one codebase; it is possible to use OpenGL and OpenGL ES shaders for an unintended use to compute highly parallel non-rendering tasks on a GPU, but it's not a good practice. OpenCL is the standard in desktop games for computation that works across GPU vendors, but the closest equivalent to OpenCL (on most Android devices) is the proprietary RenderScript language (Android-only, no port available to desktop).

I think you're right, our best hope may lie in having thread-safe Dijkstra functions that can be called by multiple CPU threads, making each creature that needs to find a path use its own thread. There aren't any static mutable variables that DijkstraMap relies on, but the same DijkstraMap cannot be used across multiple threads due to each object having a (non-static) 2D double array that it mutates very frequently.  I don't think it would be too much trouble to have a different DijkstraMap object for different creatures, though; the objects shouldn't be terribly heavyweight and enough is unique to each creature to justify a single DijkstraMap for a single monster (I recently added optional costs to enter cells, which could be useful for things like preventing fish monsters from pathfinding their way out of water just to asphyxiate, for instance).

A different, but remote, possibility would be to precompute or somehow partially precompute DijkstraMaps (possibly in parallel), compress them using some novel technique, and cache them for later lookup. The method for encoding FOV would probably be a very bad choice for DijkstraMap, and many parameters can change with pathfinding, such as suddenly having multiple goal cells, that would be constant or nonexistent in FOV, so I have doubts about this.

**Comment by tommyettinger** [comment]
Oh yes, and Hilbert Curve FOV Compression Stuff gained a bunch of new features in the last 2 days; now you can use `CoordPacker.packMulti()` on a double[][], specifying a double[] of what values in the first double[][] should be used as thresholds for different levels of (in the most common case of FOV) lighting. You can also now use `CoordPacker.unpackMultiDouble()` to decompress the short[][] that produces and get a graded double[][] out of it. You can possibly use a different levels double[] to change values meeting a threshold of 0.25 originally to 1.0 in the returned 2D array, and you could change  values originally meeting a threshold of 1.0 to 999.0 while you're at it., for instance. Yes, it's still a very good compression, absolute worst case I've tried (2D float array for FOV being turned into 8 levels of compressed curves and the map as poorly suited to a Hilbert curve as possible) is still about 4.13% the size of a float[][]. In an optimal case, with a larger map, it was closer to 0.13% the size.

There's also a small thesis worth of JavaDocs I added to CoordPacker.

**Comment by tommyettinger** [comment]
Well this issue sure got large and out of scope quickly... Since the last post, Coord has definitely seemed stable with the current immutable technique, and the relatively-unrelated CoordPacker should be usable on its own (unlikely) or in FOVCache. I think since we know that reusing Coords isn't slower than, and is much safer/somewhat more memory-efficient/probably slightly faster than the earlier approach of using mutable Points, this issue can be closed.
